src/loader.cpp

```diff
@@ -179,16 +180,44 @@ void loader_local::initialize(const json& config_json)
 
     m_provider = provider_factory::create(config_json);
 
-    unsigned int threads_num = lcfg.decode_thread_count != 0 ? lcfg.decode_thread_count
-                                                             : std::thread::hardware_concurrency();
+    // Parse list of cpu cores to set thread affinity.
+    const char*      env_cpu_list = getenv("AEON_CPU_LIST"); // Example: 0-4,30,10,32-33
+    std::vector<int> thread_affinity_map;
+    if (env_cpu_list != nullptr && std::strlen(env_cpu_list) > 0) // set affinity from environment
+    {
+        std::string str_cpu_list{env_cpu_list};
+        thread_affinity_map = nervana::parse_cpu_list(str_cpu_list);
+
+        if (!lcfg.cpu_list.empty())
+        {
+            WARN << "Both AEON_CPU_LIST and 'cpu_list' are detected. AEON_CPU_LIST environment "
+                    "variable will be used in this case.";
+        }
+    }
+    else if (!lcfg.cpu_list.empty()) // set affinity from config
+    {
+        thread_affinity_map = nervana::parse_cpu_list(lcfg.cpu_list);
+    }
+    if (thread_affinity_map.empty()) // automatically determine number of threads
+    {
+        // we don't use all threads, some of them we leave for other pipeline objects and system
+        auto nthreads =
+            std::thread::hardware_concurrency() -
+            std::min(m_max_count_of_free_threads,
+                     static_cast<int>(std::thread::hardware_concurrency() / m_free_threads_ratio));
+        thread_affinity_map.resize(nthreads);
+        std::iota(thread_affinity_map.begin(), thread_affinity_map.end(), 0);
+    }
+    unsigned int threads_num = !thread_affinity_map.empty() ? thread_affinity_map.size()
+                                                            : std::thread::hardware_concurrency();
 
     const int decode_size =
         lcfg.batch_size * ((threads_num * m_input_multiplier - 1) / lcfg.batch_size + 1);
     m_batch_iterator = make_shared<batch_iterator>(m_block_manager, decode_size);
 
     m_decoder = make_shared<batch_decoder>(m_batch_iterator,
                                            decode_size,
-                                           lcfg.decode_thread_count,
+                                           thread_affinity_map,
```

Comment:
Why not moving this `thread_affinity_map` instead of copying?

Reply:
I was hoping this would come out in a review so we can discuss a suggestion of how to optimize this. I'm not sure if we can leave it to the compiler to do a better job or explicitly implement std;:move or const reference passing or other ideas.

Reply:
ok I added some move semantics

Reply:
IMHO, const reference is the best choice, because we are not going to change that vector

Reply:
so we don't  have to use move semantics there

Reply:
What do you mean?

Reply:
thread_pool now makes a copy of it, I'm not sure if we can have thread_pool hold a const reference instead

Reply:
using r-value references or perfect forwarding made things ugly so I moved to const reference with single copy because its not a critical path or a heavy copying operation